google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketVirtualFile.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.gcs;
+
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.tools.intellij.ui.GoogleCloudToolsIcons;
+import com.intellij.openapi.fileEditor.FileEditorProvider;
+import com.intellij.openapi.fileTypes.FileType;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.openapi.vfs.VirtualFileSystem;
+import com.intellij.openapi.vfs.ex.dummy.DummyFileSystem;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import javax.swing.Icon;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A {@link VirtualFile} representation of a Google Cloud Storage bucket.
+ *
+ * <p>This is essentially just a wrapper for a {@link Bucket} needed in order to create a custom
+ * editor window for exploring GCS bucket contents. Custom editors are created by providing {@link
+ * FileEditorProvider} implementations which are backed by VirtualFiles.
+ */
+class GcsBucketVirtualFile extends VirtualFile {
```

Comment:
Can this be final?

Reply:
its being mocked in a test, so no :( (at least not without moving things around). Apparently there is an incubating feature in Mockito 2 to allow this:
https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#unmockable